### ISSUE TITLE
IllegalArgumentException  during GSON serialisation of PostStatisticalDerivedTimeSeries request

### ISSUE THREAD

**Issue by liefra** [AquaticInformatics/aquarius-sdk-java#129]
The GSON serialisation of `PostStatisticalDerivedTimeSeries` fails while trying to submit a `POST` request with the Provisioning client.

### Steps to Reproduce

````
private void createStatisticalDerivedTimeSeries() {

        Provisioning.PostStatisticalDerivedTimeSeries request =  new Provisioning.PostStatisticalDerivedTimeSeries()  ;

        request.setComputationIdentifier("Median");
        request.setComputationPeriodIdentifier("Daily");
        request.setTimeSeriesUniqueId("c32ed3b998064e01831ff4603a4fa7c4");
        request.setTimeStepCount(1);
        request.setNewValueLocation(Provisioning.NewValueLocationType.End);
        request.setLocationUniqueId("80e607fdc9414c38bf62749e28a39314");
        request.setLabel("Daily Median Test");
        request.setParameter("HG");
        request.setUtcOffset(Duration.ofHours(7));
        request.setInterpolationType(Provisioning.InterpolationType.PrecedingConstant);
        request.setMethod("HGSTAFF");
        request.setPublish(false);

        aquariusClient.Provisioning.send(request);
    }
````

### Expected Behaviour

The POST requests succeeds and replies the create time-series.

### Actual Behaviour

IllegalArgumentException thrown.

````
java.lang.IllegalArgumentException: class com.aquaticinformatics.aquarius.sdk.timeseries.servicemodels.Provisioning$PostStatisticalDerivedTimeSeries declares multiple JSON fields named ComputationIdentifier
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:170)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:100)
	at com.google.gson.Gson.getAdapter(Gson.java:423)
	at com.google.gson.Gson.toJson(Gson.java:661)
	at com.google.gson.Gson.toJson(Gson.java:648)
	at com.google.gson.Gson.toJson(Gson.java:603)
	at com.google.gson.Gson.toJson(Gson.java:583)
	at net.servicestack.client.JsonServiceClient.send(JsonServiceClient.java:319)
	at com.aquaticinformatics.aquarius.sdk.helpers.SdkServiceClient.sendRequest(SdkServiceClient.java:350)
	at net.servicestack.client.JsonServiceClient.send(JsonServiceClient.java:299)
	at com.example.AquariusDemoService.createCalc(AquariusDemoService.java:44)
	at com.example.AquariusDemoService.startUp(AquariusDemoService.java:24)
````

@

**Comment by DougSchmidt-AI** [comment]
@liefra Thanks for opening the issue. I'm reproducing that here quite easily. Hopefully it will be an easy fix.

**Comment by DougSchmidt-AI** [comment]
OK, I see the issue now. But it might take a while to fix, its been broken for a while ..

The Java service model classes are auto-generated, deriving from some .NET code that comes with the AQTS server (that's the [ServiceStack](https://github.com/ServiceStack/ServiceStack) framework we use to implement all the REST services).

The server side .NET class structure has a bad pattern that is not converting to Java properly.

Here is the (reduced) class hierarchy of the `PostStatisticalDerivedTimeSeries` request class:

```C#
[Route("/locations/{LocationUniqueId}/timeseries/statistical", HttpMethods.Post]
public class PostStatisticalDerivedTimeSeries : TimeSeriesBase
{
    [ApiMember(IsRequired = true, DataType = "string", Description = "Unique ID of the time series")]
    public Guid TimeSeriesUniqueId { get; set; }

    [ApiMember(IsRequired = true)]
    public override string ComputationIdentifier { get; set; }

    [ApiMember(IsRequired = true, Description = "Computation period identifier")]
    public override string ComputationPeriodIdentifier { get; set; }
    
    // Other properties ...
}

public abstract class TimeSeriesBase : IReturn<ResponseDtos.TimeSeries>
{
    [ApiMember(IsRequired = false)]
    public virtual string ComputationIdentifier { get; set; }

    [ApiMember(Description = "Computation period identifier")]
    public virtual string ComputationPeriodIdentifier { get; set; }
    
    // Other properties ...
}
```

That class inherits from `TimeSeriesBase`, which also defines the `ComputationIdentifier` and `ComputationPeriodIdentifier`.

The .NET code is using enough hints (the C# `virtual` keyword in the base class and the `override` keyword in the derived class) that all the JSON serialization still works.

But Java doesn't have equivalent ways of overriding base class properties in a derived class. It just actually stores two different fields, one in the base class and one the derived class, and which field (or method) is used depends on the source code context. Yuck!

In this case, the GSON serialization library is using reflection to figure out which fields to serialize into JSON, and it finds two fields with the same name and throws an exception.

I'll need to dig deep into the GSON code base and the ServiceStack Java code generator to see if there is some sort of workaround. I will keep you posted on what I find.

**Comment by DougSchmidt-AI** [comment]
@liefra I've inspected the code for other instances of this .NET-only `virtual`/`override` pattern. There are just two affected request DTOs.

- `PostStatisticalDerivedTimeSeries` (the `ComputationIdentifier` and `ComputationPeriodIdentifier` properties, which you already know about)
- `PutQualityCode` (the `GradeCode` property also uses this pattern)

All the other request/response DTOs, across the Publish, Acquisition, and Provisioning APIs will work in the Java SDK.

**Comment by DougSchmidt-AI** [comment]
Here is a workaround using the current (20.2.2) SDK.

I've flattenned the two inherited classes into a single request DTO class, avoiding the duplicate field names.

If you paste this class into your integration, then you can create the stats time-series using this class, until we officially fix it in a subsequent release.

```java
    @Route(Path="/locations/{LocationUniqueId}/timeseries/statistical", Verbs="POST")
    public static class HackedPostStatisticalDerivedTimeSeries implements IReturn<Provisioning.TimeSeries>
    {
        /**
         * Unique ID of the location for which a time series is to be created
         */
        @ApiMember(DataType="string", Description="Unique ID of the location for which a time series is to be created", IsRequired=true)
        public String LocationUniqueId = null;

        /**
         * Label
         */
        @ApiMember(Description="Label", IsRequired=true)
        public String Label = null;

        /**
         * The ID of the parameter
         */
        @ApiMember(Description="The ID of the parameter", IsRequired=true)
        public String Parameter = null;

        /**
         * The ID of the unit
         */
        @ApiMember(Description="The ID of the unit", IsRequired=true)
        public String Unit = null;

        /**
         * Interpolation type
         */
        @ApiMember(DataType="InterpolationType", Description="Interpolation type", IsRequired=true)
        public Provisioning.InterpolationType InterpolationType = null;

        /**
         * Sub location identifier
         */
        @ApiMember(Description="Sub location identifier")
        public String SubLocationIdentifier = null;

        /**
         * ISO 8601 Duration Format
         */
        @ApiMember(DataType="Offset", Description="ISO 8601 Duration Format")
        public Duration UtcOffset = null;

        /**
         * Publish. If set true, will enforce that Location Publish is also true
         */
        @ApiMember(DataType="boolean", Description="Publish. If set true, will enforce that Location Publish is also true")
        public Boolean Publish = null;

        /**
         * Description
         */
        @ApiMember(Description="Description")
        public String Description = null;

        /**
         * Comment
         */
        @ApiMember(Description="Comment")
        public String Comment = null;

        /**
         * Monitoring method code
         */
        @ApiMember(Description="Monitoring method code", IsRequired=true)
        public String Method = null;

        /**
         * Extended attribute values
         */
        @ApiMember(DataType="Array<ExtendedAttributeValue>", Description="Extended attribute values")
        public List<Provisioning.ExtendedAttributeValue> ExtendedAttributeValues = null;

        /**
         * Unique ID of the time series
         */
        @ApiMember(DataType="string", Description="Unique ID of the time series", IsRequired=true)
        public String TimeSeriesUniqueId = null;

        @ApiMember(IsRequired=true)
        public String ComputationIdentifier = null;

        /**
         * Computation period identifier
         */
        @ApiMember(Description="Computation period identifier", IsRequired=true)
        public String ComputationPeriodIdentifier = null;

        /**
         * New value location
         */
        @ApiMember(DataType="NewValueLocationType", Description="New value location")
        public Provisioning.NewValueLocationType NewValueLocation = null;

        /**
         * Require minimum coverage
         */
        @ApiMember(DataType="boolean", Description="Require minimum coverage")
        public Boolean RequireMinimumCoverage = null;

        /**
         * Coverage minimum percentage
         */
        @ApiMember(DataType="double", Description="Coverage minimum percentage")
        public Double CoverageMinimumPercentage = null;

        /**
         * Partial coverage grade
         */
        @ApiMember(DataType="integer", Description="Partial coverage grade")
        public Integer PartialCoverageGrade = null;

        /**
         * Observation offset in minutes
         */
        @ApiMember(DataType="integer", Description="Observation offset in minutes")
        public Integer ObservationOffsetInMinutes = null;

        /**
         * Time Step Count. Must be included for 'Statistic' derived time-series.
         */
        @ApiMember(DataType="integer", Description="Time Step Count. Must be included for 'Statistic' derived time-series.")
        public Integer TimeStepCount = null;

        public String getLocationUniqueId() { return LocationUniqueId; }
        public HackedPostStatisticalDerivedTimeSeries setLocationUniqueId(String value) { this.LocationUniqueId = value; return this; }
        public String getLabel() { return Label; }
        public HackedPostStatisticalDerivedTimeSeries setLabel(String value) { this.Label = value; return this; }
        public String getParameter() { return Parameter; }
        public HackedPostStatisticalDerivedTimeSeries setParameter(String value) { this.Parameter = value; return this; }
        public String getUnit() { return Unit; }
        public HackedPostStatisticalDerivedTimeSeries setUnit(String value) { this.Unit = value; return this; }
        public Provisioning.InterpolationType getInterpolationType() { return InterpolationType; }
        public HackedPostStatisticalDerivedTimeSeries setInterpolationType(Provisioning.InterpolationType value) { this.InterpolationType = value; return this; }
        public String getSubLocationIdentifier() { return SubLocationIdentifier; }
        public HackedPostStatisticalDerivedTimeSeries setSubLocationIdentifier(String value) { this.SubLocationIdentifier = value; return this; }
        public Duration getUtcOffset() { return UtcOffset; }
        public HackedPostStatisticalDerivedTimeSeries setUtcOffset(Duration value) { this.UtcOffset = value; return this; }
        public Boolean isPublish() { return Publish; }
        public HackedPostStatisticalDerivedTimeSeries setPublish(Boolean value) { this.Publish = value; return this; }
        public String getDescription() { return Description; }
        public HackedPostStatisticalDerivedTimeSeries setDescription(String value) { this.Description = value; return this; }
        public String getComment() { return Comment; }
        public HackedPostStatisticalDerivedTimeSeries setComment(String value) { this.Comment = value; return this; }
        public String getMethod() { return Method; }
        public HackedPostStatisticalDerivedTimeSeries setMethod(String value) { this.Method = value; return this; }
        public List<Provisioning.ExtendedAttributeValue> getExtendedAttributeValues() { return ExtendedAttributeValues; }
        public HackedPostStatisticalDerivedTimeSeries setExtendedAttributeValues(List<Provisioning.ExtendedAttributeValue> value) { this.ExtendedAttributeValues = value; return this; }
        public String getTimeSeriesUniqueId() { return TimeSeriesUniqueId; }
        public HackedPostStatisticalDerivedTimeSeries setTimeSeriesUniqueId(String value) { this.TimeSeriesUniqueId = value; return this; }
        public String getComputationIdentifier() { return ComputationIdentifier; }
        public HackedPostStatisticalDerivedTimeSeries setComputationIdentifier(String value) { this.ComputationIdentifier = value; return this; }
        public String getComputationPeriodIdentifier() { return ComputationPeriodIdentifier; }
        public HackedPostStatisticalDerivedTimeSeries setComputationPeriodIdentifier(String value) { this.ComputationPeriodIdentifier = value; return this; }
        public Provisioning.NewValueLocationType getNewValueLocation() { return NewValueLocation; }
        public HackedPostStatisticalDerivedTimeSeries setNewValueLocation(Provisioning.NewValueLocationType value) { this.NewValueLocation = value; return this; }
        public Boolean isRequireMinimumCoverage() { return RequireMinimumCoverage; }
        public HackedPostStatisticalDerivedTimeSeries setRequireMinimumCoverage(Boolean value) { this.RequireMinimumCoverage = value; return this; }
        public Double getCoverageMinimumPercentage() { return CoverageMinimumPercentage; }
        public HackedPostStatisticalDerivedTimeSeries setCoverageMinimumPercentage(Double value) { this.CoverageMinimumPercentage = value; return this; }
        public Integer getPartialCoverageGrade() { return PartialCoverageGrade; }
        public HackedPostStatisticalDerivedTimeSeries setPartialCoverageGrade(Integer value) { this.PartialCoverageGrade = value; return this; }
        public Integer getObservationOffsetInMinutes() { return ObservationOffsetInMinutes; }
        public HackedPostStatisticalDerivedTimeSeries setObservationOffsetInMinutes(Integer value) { this.ObservationOffsetInMinutes = value; return this; }
        public Integer getTimeStepCount() { return TimeStepCount; }
        public HackedPostStatisticalDerivedTimeSeries setTimeStepCount(Integer value) { this.TimeStepCount = value; return this; }
        private static Object responseType = Provisioning.TimeSeries.class;
        public Object getResponseType() { return responseType; }
    }
```

Then just change your call site to use the new "hacked" version, and you should be good to go.

```java
HackedPostStatisticalDerivedTimeSeries request =  new HackedPostStatisticalDerivedTimeSeries();

// ... Set all your properties ...

// Also, don't forget to set the unit of the time-series
request.setUnit("m");

aquariusClient.Provisioning.send(request);
```

**Comment by liefra** [comment]
@DougSchmidt-AI : Thank you for have a look at this issue and for providing a workaround. It is good to know that only one other request DTO `PutQualityCode` is affected and all other work well. I'm not familiar with the .NET-only `virtual/override` pattern and was looking so far at a way to tweak the GSON serialisation configuration to get rid of this issue. 

But now, I tested the workaround and it works :) Thanks

**Comment by DougSchmidt-AI** [comment]
Glad to hear you've got it working.

I've added an internal ticket to our backlog, so we can solve this more robustly in future releases. I'll leave this GitHub issue open, and resolve it once the hack is no longer needed.

And while I try to keep on top of the recently-openned GitHub issues, this one fell through the crack of my GMail filter (I get hundreds of GitHub emails a day). I just happenned to catch it by chance. So next time, if you find an issue you've raised not getting any response, feel free to open a ticket in our [Support Portal](https://aquaticinformatics.deskpro.com/) if your organization has an account, since those API/SDK questions get routed quite quickly to me.

Cheers,
Doug

**Comment by DougSchmidt-AI** [comment]
@liefra I've made a fix to the core TimeSeries code, so that the .NET `virtual/override` pattern is no longer used. This fix will land in the AQTS 2020.3 release,  currently scheduled for late October. When 2020.3 is released, I'll update the Java SDK and the problem should be fixed and I can finally close this issue.

**Comment by DougSchmidt-AI** [comment]
@liefra SDK 20.3.0 has been released to the Maven Central repository, with this fix. So I can finally close this issue.

https://search.maven.org/search?q=aquarius.sdk
